File: src/model_and_state_demo.cpp
//
// Created by yongxi on 2021/5/26.
//

#include "robot_model/KinematicsLoader.h"
#include "robot_model/RobotModelLoader.h"
#include "moveit/robot_state/robot_state.h"
#include <iostream>
#include "kdl_parser/kdl_parser.hpp"
#include "moveit/robot_model/robot_model.h"

int main() {

    // Create robot model
    RobotModelLoaderPtr loader = createRobotModelLoaderFromFile("resources/panda_arm_hand.urdf",
                                                                "resources/panda_arm_hand.srdf");
    std::cout << "Model Ref use after creation: " << loader->countModelUse() << std::endl;
    // Print out info of robot model
    std::cout << "===== RobotModel Info =====" << std::endl;
    //loader->outputModelInfo(std::cout);

    // Print out current joint limits info
    std::cout << "\n===== Joint Limits Info =====" << std::endl;
    //loader->outputLimitsInfo(std::cout);

    // Allocate Kinematics Solver
    /*
    std::cout << "\n===== Allocate Kinematics Solver =====" << std::endl;
    loader.initKinematicFromFile("resources/kinematics.yaml");
    moveit::core::JointModelGroup* jmg = loader.getRobotModel()->getJointModelGroup("panda_arm");
    kinematics::KinematicsBasePtr solver = loader.allocateSolver(jmg);
    std::cout << "Kinematics solver created for group " << solver->getGroupName() << std::endl;
    std::cout << "From " << solver->getBaseFrame() << " To " << solver->getTipFrame() << std::endl;
    */

    // Implement MoveIt robot model and robot state tutorial.
    std::cout << "\n===== Implement MoveIt robot model and robot state tutorial =====" << std::endl;
    robot_model::RobotModelPtr kinematic_model = loader->getRobotModel();
    moveit::core::JointModelGroup* jmg = loader->getRobotModel()->getJointModelGroup("panda_arm");
    // rdf_loader::RDFLoaderPtr rdf_loader_ = loader->getRDFLoader();
    KinematicsLoaderPtr kinematics_loader = createKinematicsLoaderFromFile("resources/kinematics.yaml");
    //kinematic_model->setKinematicsAllocators(std::map<std::string, moveit::core::SolverAllocatorFn>{
    //        {"panda_arm", kinematics_loader->getAllocatorFn(rdf_loader_)}
    //});
    loader->loadKinematicsSolvers(kinematics_loader);

    //std::cout << "Model frame: " << kinematic_model->getModelFrame() << std::endl;
    std::printf("Model frame: %s\n", kinematic_model->getModelFrame().c_str());
    //robot_state::RobotStatePtr kinematic_state(new robot_state::RobotState(kinematic_model));
    std::cout << "model ptr use before state: " << loader->countModelUse() << std::endl;
    robot_state::RobotStatePtr kinematic_state = loader->newRobotState();
    std::cout << "model ptr use after state: " << loader->countModelUse() << std::endl;
    kinematic_state->setToDefaultValues();

    const std::vector<std::string>& joint_names = jmg->getVariableNames();
    std::vector<double> joint_values;
    kinematic_state->copyJointGroupPositions(jmg, joint_values);
    std::printf("\n== Get Joint Value ==\n");
    std::printf("Current position:\n");
    for (std::size_t i = 0; i< joint_names.size(); ++i) {
        std::printf("\tJoint %s: %f\n", joint_names[i].c_str(), joint_values[i]);
    }

    std::printf("\n== Joint Limit ==\n");
    std::printf("Set position out of bound\n");
    joint_values[0] = 5.57;
    kinematic_state ->setJointGroupPositions(jmg, joint_values);
    std::cout << "Current state is " << (kinematic_state->satisfiesBounds() ? "valid" : "not valid") << std::endl;

    std::printf("Enforce bounds\n");
    kinematic_state->enforceBounds();
    std::cout << "Current state is " << (kinematic_state->satisfiesBounds() ? "valid" : "not valid") << std::endl;

    std::printf("\n== Forward Kinematics ==\n");
    kinematic_state->setToRandomPositions(jmg);
    const Eigen::Affine3d& end_effector_state = kinematic_state ->getGlobalLinkTransform("panda_link8");
    std::cout << "Translation: \n" << end_effector_state.translation() << std::endl;
    std::cout << "Rotation: \n" << end_effector_state.rotation() << std::endl;

    std::printf("\n== Inverse Kinematics ==\n");
    bool found_ik = kinematic_state->setFromIK(jmg, end_effector_state, 10, 0.1);
    if(found_ik) {
        kinematic_state ->copyJointGroupPositions(jmg, joint_values);
        for (std::size_t i = 0; i< joint_names.size(); ++i) {
            std::printf("\tJoint %s: %f\n", joint_names[i].c_str(), joint_values[i]);
        }
    } else {
        std::cout << "Did not find IK solution" << std::endl;
    }

    std::printf("\n== Get the Jacobian ==\n");
    Eigen::Vector3d reference_point_position(0.0, 0.0, 0.0);
    Eigen::MatrixXd jacobian;
    kinematic_state ->getJacobian(jmg,
                                  kinematic_state->getLinkModel(jmg->getLinkModelNames().back()),
                                  reference_point_position, jacobian);

    std::cout << "Jacobian:\n" << jacobian << std::endl;


    std::cout << "sizeof(double) = " << sizeof(double) << std::endl;
    return 0;
}

File: example/model_and_state_demo.py
import sys
sys.path.append("../install/lib")
import numpy
import moveit_noros as moveit

print("==== Implement Robot Model and Robot State Tutorial From Moveit Tutorials. ====")
print("== Load robot model ==")
robot_loader = moveit.createRobotModelLoaderFromFile("../resources/franka_convert.urdf", "../resources/panda_arm_hand.srdf")
kinematic_solver_loader = moveit.createKinematicsLoaderFromFile("../resources/kinematics.yaml")
robot_loader.loadKinematicsSolvers(kinematic_solver_loader)
kinematic_model = robot_loader.getModel()
print("Model Frame: %s" % kinematic_model.getModelFrame())

PLANNING_GROUP = "panda_arm"
kinematic_state = robot_loader.newRobotState()
kinematic_state.setToDefaultValues()

joint_model_group = kinematic_model.getJointModelGroup(PLANNING_GROUP)
joint_names = joint_model_group.getVariableNames()
joint_values = kinematic_state.copyJointGroupPositions(joint_model_group)
print("== Current joint values ==")
for i in range(len(joint_names)):
    print("\t%s:\t%f" % (joint_names[i], joint_values[i]))

print("== Test Joint Limits ==")
print("Set joint value out of bound.")
joint_values=joint_values=[-1.3177302367227017, -1.7353981633974485, 0.8012309797768007, -2.295736269770167, 1.9069654787403831, 2.296614604292792, 2.5637651905206544]
kinematic_state.setJointGroupPositions(joint_model_group, joint_values)
if kinematic_state.satisfiesBounds():
    print("Current state is valid.")
else:
    print("Current state is invalid.")
print("Enforce robot state bounds.")
kinematic_state.enforceBounds()
if kinematic_state.satisfiesBounds():
    print("Current state is valid.")
else:
    print("Current state is invalid.")

print("== Forward Kinematics ==")
print("Set to random position")
#kinematic_state.setToRandomPositions(joint_model_group)
end_effector_state = kinematic_state.getGlobalLinkTransform("panda_link7")
print(end_effector_state)
#print("Translation:")
#print(end_effector_state.translation())
#print("Rotation:")
#print(end_effector_state.rotation())


print("== Inverse Kinematics ==")
found_ik = kinematic_state.setFromIK(joint_model_group, end_effector_state)
if found_ik:
    joint_values = kinematic_state.copyJointGroupPositions(joint_model_group)
    for i in range(len(joint_names)):
        print("\t%s:\t%f" % (joint_names[i], joint_values[i]))
else:
    print("Did not find IK solution")

print("== Get the Jacobian ==")
reference_point_position = numpy.array([0.0, 0.0, 0.0], dtype='double')
print(kinematic_state.getJacobian(joint_model_group, joint_model_group.getLinkModelNames()[-1], reference_point_position))
